### ISSUE TITLE
PaddleNLP/examples/text_classification/pretrained_models/train.py 评价的时候如何输出 精确率、召回率和F1值

### ISSUE THREAD

**Issue by AI-Mart** [PaddlePaddle/PaddleNLP#1183]
以下函数，目前只有准确率，评价的时候如何增加输出 精确率、召回率和F1值，具体增加什么代码？

```python
@paddle.no_grad()
def evaluate(model, criterion, metric, data_loader):
    """
    Given a dataset, it evals model and computes the metric.

    Args:
        model(obj:`paddle.nn.Layer`): A model to classify texts.
        data_loader(obj:`paddle.io.DataLoader`): The dataset loader which generates batches.
        criterion(obj:`paddle.nn.Layer`): It can compute the loss.
        metric(obj:`paddle.metric.Metric`): The evaluation metric.
    """
    model.eval()
    metric.reset()
    losses = []
    for batch in data_loader:
        input_ids, token_type_ids, labels = batch
        logits = model(input_ids, token_type_ids)
        loss = criterion(logits, labels)
        losses.append(loss.numpy())
        correct = metric.compute(logits, labels)
        metric.update(correct)
        accu = metric.accumulate()
    print("eval loss: %.5f, accu: %.5f" % (np.mean(losses), accu))
    model.train()
    metric.reset()
```

**Comment by linjieccc** [comment]
你好，可以使用PaddleNLP内置Metrics API中的`paddlenlp.metrics.AccuracyAndF1`，具体示例如下

metric替换为`paddlenlp.metrics.AccuracyAndF1`：
```python
from paddlenlp.metrics import AccuracyAndF1
metric = AccyractAndF1()
```

`evaluate`函数修改如下：
```python
@paddle.no_grad()
def evaluate(model, criterion, metric, data_loader):
    """
    Given a dataset, it evals model and computes the metric.

    Args:
        model(obj:`paddle.nn.Layer`): A model to classify texts.
        data_loader(obj:`paddle.io.DataLoader`): The dataset loader which generates batches.
        criterion(obj:`paddle.nn.Layer`): It can compute the loss.
        metric(obj:`paddle.metric.Metric`): The evaluation metric.
    """
    model.eval()
    metric.reset()
    losses = []
    for batch in data_loader:
        input_ids, token_type_ids, labels = batch
        logits = model(input_ids, token_type_ids)
        loss = criterion(logits, labels)
        losses.append(loss.numpy())
        correct = metric.compute(logits, labels)
        metric.update(correct)
    res = metric.accumulate()
    accu, precision, recall, f1 = res[0], res[1], res[2], res[3] 
    print("eval loss: %.5f, accu: %.5f, precision: %.5f, recall: %.5f, f1: %.5f" % (np.mean(losses), accu, precision, recall, f1))
    model.train()
    metric.reset()
```

**Comment by AI-Mart** [comment]
，谢谢大佬， 刚才挨个搜了下，找到glue的metric做参考，也是差不多这么写的，已经在跑了
有个疑问的是res = metric.accumulate()是不是要放到for循环外面，做评估是针对所有的评估数据然后取平均吧，放里面是每个batch一个评估


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PaddlePaddle/PaddleNLP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月18日(星期一) 中午11:45
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PaddlePaddle/PaddleNLP] PaddleNLP/examples/text_classification/pretrained_models/train.py 评价的时候如何输出 精确率、召回率和F1值 (Issue #1183)





 
你好，可以使用PaddleNLP内置Metrics API中的paddlenlp.metrics.AccuracyAndF1，具体示例如下
 
metric替换为paddlenlp.metrics.AccuracyAndF1：
 from paddlenlp.metrics import AccuracyAndF1 metric = AccyractAndF1()
 
evalution函数修改如下：
 @paddle.no_grad() def evaluate(model, criterion, metric, data_loader):     """     Given a dataset, it evals model and computes the metric.      Args:         model(obj:`paddle.nn.Layer`): A model to classify texts.         data_loader(obj:`paddle.io.DataLoader`): The dataset loader which generates batches.         criterion(obj:`paddle.nn.Layer`): It can compute the loss.         metric(obj:`paddle.metric.Metric`): The evaluation metric.     """     model.eval()     metric.reset()     losses = []     for batch in data_loader:         input_ids, token_type_ids, labels = batch         logits = model(input_ids, token_type_ids)         loss = criterion(logits, labels)         losses.append(loss.numpy())         correct = metric.compute(logits, labels)         metric.update(correct)         res = metric.accumulate()         accu, precision, recall, f1 = res[0], res[1], res[2], res[3]      print("eval loss: %.5f, accu: %.5f, precision: %.5f, recall: %.5f, f1: %.5f" % (np.mean(losses), accu, precision, recall, f1))     model.train()     metric.reset()
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by AI-Mart** [comment]
另外一个问题假如要作为保存最佳模型的判别指标，是不是针对2分类问题，F1指标是最佳判断指标




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Lincoln"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月18日(星期一) 中午11:50
***@***.******@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;回复： [PaddlePaddle/PaddleNLP] PaddleNLP/examples/text_classification/pretrained_models/train.py 评价的时候如何输出 精确率、召回率和F1值 (Issue #1183)



，谢谢大佬， 刚才挨个搜了下，找到glue的metric做参考，也是差不多这么写的，已经在跑了
有个疑问的是res = metric.accumulate()是不是要放到for循环外面，做评估是针对所有的评估数据然后取平均吧，放里面是每个batch一个评估


------------------ 原始邮件 ------------------
发件人:                                                                                                                        "PaddlePaddle/PaddleNLP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月18日(星期一) 中午11:45
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PaddlePaddle/PaddleNLP] PaddleNLP/examples/text_classification/pretrained_models/train.py 评价的时候如何输出 精确率、召回率和F1值 (Issue #1183)





 
你好，可以使用PaddleNLP内置Metrics API中的paddlenlp.metrics.AccuracyAndF1，具体示例如下
 
metric替换为paddlenlp.metrics.AccuracyAndF1：
 from paddlenlp.metrics import AccuracyAndF1 metric = AccyractAndF1()
 
evalution函数修改如下：
 @paddle.no_grad() def evaluate(model, criterion, metric, data_loader):     """     Given a dataset, it evals model and computes the metric.      Args:         model(obj:`paddle.nn.Layer`): A model to classify texts.         data_loader(obj:`paddle.io.DataLoader`): The dataset loader which generates batches.         criterion(obj:`paddle.nn.Layer`): It can compute the loss.         metric(obj:`paddle.metric.Metric`): The evaluation metric.     """     model.eval()     metric.reset()     losses = []     for batch in data_loader:         input_ids, token_type_ids, labels = batch         logits = model(input_ids, token_type_ids)         loss = criterion(logits, labels)         losses.append(loss.numpy())         correct = metric.compute(logits, labels)         metric.update(correct)         res = metric.accumulate()         accu, precision, recall, f1 = res[0], res[1], res[2], res[3]      print("eval loss: %.5f, accu: %.5f, precision: %.5f, recall: %.5f, f1: %.5f" % (np.mean(losses), accu, precision, recall, f1))     model.train()     metric.reset()
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by linjieccc** [comment]
Accuracy在分类问题中是评估模型好坏的经典评估指标，但它对数据集的样本分布要求较高。

如果是在数据集分布不均衡等情况下可以选择F1值（trade off Precision VS Recall）作为模型好坏的评判标准。

**Comment by linjieccc** [comment]
感谢指出，我们修复一下，谢谢！

**Comment by linjieccc** [comment]
> ，谢谢大佬， 刚才挨个搜了下，找到glue的metric做参考，也是差不多这么写的，已经在跑了 有个疑问的是res = metric.accumulate()是不是要放到for循环外面，做评估是针对所有的评估数据然后取平均吧，放里面是每个batch一个评估
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "PaddlePaddle/PaddleNLP" ***@***.***&gt;; 发送时间:&nbsp;2021年10月18日(星期一) 中午11:45 ***@***.***&gt;; ***@***.******@***.***&gt;; 主题:&nbsp;Re: [PaddlePaddle/PaddleNLP] PaddleNLP/examples/text_classification/pretrained_models/train.py 评价的时候如何输出 精确率、召回率和F1值 (Issue #1183) 你好，可以使用PaddleNLP内置Metrics API中的paddlenlp.metrics.AccuracyAndF1，具体示例如下 metric替换为paddlenlp.metrics.AccuracyAndF1： from paddlenlp.metrics import AccuracyAndF1 metric = AccyractAndF1() evalution函数修改如下： @paddle.no_grad() def evaluate(model, criterion, metric, data_loader): """ Given a dataset, it evals model and computes the metric. Args: model(obj:`paddle.nn.Layer`): A model to classify texts. data_loader(obj:`paddle.io.DataLoader`): The dataset loader which generates batches. criterion(obj:`paddle.nn.Layer`): It can compute the loss. metric(obj:`paddle.metric.Metric`): The evaluation metric. """ model.eval() metric.reset() losses = [] for batch in data_loader: input_ids, token_type_ids, labels = batch logits = model(input_ids, token_type_ids) loss = criterion(logits, labels) losses.append(loss.numpy()) correct = metric.compute(logits, labels) metric.update(correct) res = metric.accumulate() accu, precision, recall, f1 = res[0], res[1], res[2], res[3] print("eval loss: %.5f, accu: %.5f, precision: %.5f, recall: %.5f, f1: %.5f" % (np.mean(losses), accu, precision, recall, f1)) model.train() metric.reset() — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe. Triage notifications on the go with GitHub Mobile for iOS or Android.

感谢指出，我们修复下，谢谢！

**Comment by AI-Mart** [comment]
好的，谢谢，样本情况有些确实未知，那还是选F1保险




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PaddlePaddle/PaddleNLP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月18日(星期一) 下午2:39
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PaddlePaddle/PaddleNLP] PaddleNLP/examples/text_classification/pretrained_models/train.py 评价的时候如何输出 精确率、召回率和F1值 (Issue #1183)





 
Accuracy在分类问题中是评估模型好坏的经典评估指标，但它对数据集的样本分布要求较高。
 
如果是在数据集分布不均衡等情况下可以选择F1值（trade off Precision VS Recall）作为模型好坏的评判标准。
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by AI-Mart** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PaddlePaddle/PaddleNLP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月18日(星期一) 下午2:40
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PaddlePaddle/PaddleNLP] PaddleNLP/examples/text_classification/pretrained_models/train.py 评价的时候如何输出 精确率、召回率和F1值 (Issue #1183)





 
感谢指出，我们修复一下，谢谢！
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
